devlib/utils/ssh.py

```diff
@@ -367,25 +367,32 @@ def __init__(self,
         else:
             logger.debug('Using SFTP for file transfer')
 
-        self.client = self._make_client()
-        atexit.register(self.close)
-
-        # Use a marker in the output so that we will be able to differentiate
-        # target connection issues with "password needed".
-        # Also, sudo might not be installed at all on the target (but
-        # everything will work as long as we login as root). If sudo is still
-        # needed, it will explode when someone tries to use it. After all, the
-        # user might not be interested in being root at all.
-        self._sudo_needs_password = (
-            'NEED_PASSWORD' in
-            self.execute(
-                # sudo -n is broken on some versions on MacOSX, revisit that if
-                # someone ever cares
-                'sudo -n true || echo NEED_PASSWORD',
-                as_root=False,
-                check_exit_code=False,
+        self.client = None
+        try:
+            self.client = self._make_client()
+            atexit.register(self.close)
+
+            # Use a marker in the output so that we will be able to differentiate
+            # target connection issues with "password needed".
+            # Also, sudo might not be installed at all on the target (but
+            # everything will work as long as we login as root). If sudo is still
+            # needed, it will explode when someone tries to use it. After all, the
+            # user might not be interested in being root at all.
+            self._sudo_needs_password = (
+                'NEED_PASSWORD' in
+                self.execute(
+                    # sudo -n is broken on some versions on MacOSX, revisit that if
+                    # someone ever cares
+                    'sudo -n true || echo NEED_PASSWORD',
+                    as_root=False,
+                    check_exit_code=False,
+                )
             )
-        )
+
+        except BaseException:
```

Comment:
This is one of the cases where `BaseException` is appropriate: we want to cleanup the situation as far as we can in case we get interrupted, even if the user asked for the interruption. Since we re-raise the exception, we are not going to accidentally mask it. The only bad thing I could see happen is this code being called from a loop that retries in case of failures. If `__init__` failed because the user used ctrl-c, leading to a broken state in the client, leading to client.close() to raise an exception. Then this exception would propagate, and the retry logic would swallow it and retry. The fix would be ensuring the initial exception is always preserved:
```
        except BaseException as e:
            try:
                if self.client is not None:
                    self.client.close()
            except BaseException:
                raise e
```

Reply:
Included this fix with pylint enhancements.

Reply:
This is a bug that needs to be fixed separately. When `__init__` is called, the object is already created, as `__init__` is not a constructor per say, it only initializes an already-constructed object. That means that even if `__init__` ends up raising, `__del__` will still be called. So `__del__` needs to be prepared to deal with an object where `__init__` never ran (or e.g. raised on the first line before setting any attribute).

I'm not quite sure why that problem is masked by your initial code though, but I suppose it doesn't really matter

Reply:
EDIT: actually there is an issue with my code, it should be:
```python
        except BaseException as e:
            try:
                if self.client is not None:
                    self.client.close()
            finally:
                raise e
```
The reason you saw that issue is because the exception was discarded and never re-raised in the self.client.close() path. This means that the metaprogramming `__del__` relies on told it that the object finished initializing cleanly, but in reality  it did not and was missing an attribute.

Reply:
Applied this patch.